AEROGEAR-2174 Initial interfaces and classes for SecurityService

aidenkeating opened the pull request:
## Motivation

Need initial interfaces to start development on implementation.

JIRA: https://issues.jboss.org/browse/AEROGEAR-2174

## Description

Add initial interfaces and classes

## Progress

- [x] Interfaces and classes
- [x] JavaDoc
- [x] Tests

aidenkeating commented:
@wei-lee @ciaranRoche @rachael-oregan Would you mind taking a look? This is based on the comment on this Gist (https://gist.github.com/secondsun/2e674103c85363f0a602011d44b6b44a). Still requires some work but I wanted to get the concept reviewed quickly in case it's a no go.

This allows for chaining using the SecurityService.chain().method().method().run() syntax and singular tests using the SecurityService.method() syntax. There's a bit of repetition in repeating the method names though.

ziccardi commented:
Wouldn't be better to have something like this https://gist.github.com/ziccardi/5b7a4f8e1c4394b0f92ee9a5ae5450b2 ?

That is a kind of `command pattern`. That way adding and chaining new checks can be very easy and we won't need to add a new method to add a new check.

Moreover, if I didn't misunderstood, that would be closer to what has been discussed in https://issues.jboss.org/browse/AEROGEAR-2174.

@wei-lee @secondsun @aidenkeating WDYT?

rachael-oregan commented:
The approach so far looks good to me

wei-lee commented:
@ziccardi If you look at this gist: https://gist.github.com/secondsun/2e674103c85363f0a602011d44b6b44a.

I added a comment there and I think it will be good the have methods that explicitly state what it does. With the help of autocompletion it will be very easy for developers to find out what checks are available. Yes we will need to define an extra method but I am not too concerned about it.

When implementing the checks, I think we will certain follow the `command pattern`.

ziccardi commented:
@wei-lee I'm usually against classes with too many methods. If the only benefit we get is just for code completion, an Enum could do the trick as well. I updated my gist https://gist.github.com/ziccardi/5b7a4f8e1c4394b0f92ee9a5ae5450b2.

In that way, we would have code completion and chaining. Moreover, the user will be able to add custom security checks.

aidenkeating commented:
@ziccardi @wei-lee Made another branch based on that Gist (https://github.com/aerogear/aerogear-android-sdk/compare/master...aidenkeating:AEROGEAR-2174-alt). To me the approach in the Gist seems quite nice. So I'd be happy to go with that approach if you all are? (It also includes the rooted check as an example)

@rachael-oregan @ciaranRoche ^^

wei-lee commented:
@aidenkeating yes, I think it's quite nice. Nice work and good collaboration @aidenkeating @ziccardi

rachael-oregan commented:
+1 on the alternative approach.  It's that bit neater.  Nice work!

aidenkeating commented:
@wei-lee Based on this comment https://gist.github.com/secondsun/2e674103c85363f0a602011d44b6b44a . Is this kind of style what we want for `sendMetrics`?

We could also retrieve the metrics service from `MobileCore` inside the module instead of asking the user to provide it, or have both as an option.

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,73 @@
+import java.util.ArrayList;
+
+public class SecurityCheckExecutorImpl implements SecurityCheckExecutor {
+    private final ArrayList<SecurityCheck> checks;
```
Can we have the same check more than one time? If no, this should be a Set.
Moreover, it is a good practice to never declare a variable as ArrayList: better to declare it as List.
In our case, I think declaring it as Collection would be the best choice: that way we are able to change the real container without touching the rest of the code.

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public SecurityCheckExecutor addCheck(Check check) {
```
I think that we should add 

`public SecurityCheckExecutor addCheck(SecurityCheck check) {`

too. That way the user will be able to add custom checks as well

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,73 @@
+     * @return Array of results.
+     */
+    private SecurityCheckResult[] getTestResults() {
+        ArrayList<SecurityCheckResult> results = new ArrayList<>();
```
Again here, better declare 
`List<SecurityCheckResult> results = new ArrayList<>();`

matzew replied:
and make it final :)

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,73 @@
+    private SecurityCheckResult[] getTestResults() {
+        ArrayList<SecurityCheckResult> results = new ArrayList<>();
+        for (SecurityCheck check : checks) {
+            results.add(check.test(context));
```
Do we execute all the tests synchronously?

aidenkeating replied:
Yep we do

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckResultImpl.java:
```diff
@@ -0,0 +1,24 @@
+    private final String name;
+    private final boolean passed;
+
+    public SecurityCheckResultImpl(String name, boolean passed) {
```
Make parameter final

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/metrics/SecurityCheckResultMetric.java:
```diff
@@ -0,0 +1,34 @@
+    private final String identifier;
+    private final Map<String, String> data;
+
+    public SecurityCheckResultMetric(SecurityCheckResult result) {
```
Make parameter final

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/metrics/SecurityCheckResultMetric.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Override
+    public Map<String, String> data() {
+        return data;
```
Never return reference to mutable internals.
Change this to:
`return Collections.unmodifiableMap(Data)`

ziccardi commented on core/src/main/java/org/aerogear/mobile/core/MobileCore.java:
```diff
@@ -145,7 +145,7 @@ public void destroy() {
 
             if (serviceConfiguration == null) {
-                serviceConfiguration = getServiceConfiguration(serviceModule.type());
+                serviceConfiguration = getServiceConfiguration(serviceModule.type(), serviceModule.requiresConfiguration());
```
I think it would be better to leave the 
`getServiceConfiguration(serviceModule.type())` as it is, without the `requiresConfiguration` option and simply making it returns null if no configuration exists.
Then here we could do:
`if (config == null && serviceModule.requiresConfiguration()) {
    throw new ConfigurationNotFoundException(...)`
WDYT?

aidenkeating replied:
I'm happy to allow the service configuration to be null and have each service handle what happens in that scenario. @danielpassos @wei-lee Would you be fine with that?

ziccardi replied:
@aidenkeating that's not exactly what I said. Only the service knows if they need or not a configuration, however, if they need it, it is managed centrally inside the `getInstance` method of the `MobileCore`. From the outside the result is the same, but IMO the code is cleaner this way (because from the POV of the `getConfiguration` method, not finding a configuration is not an exceptional case, so no exception should be raised).

aidenkeating replied:
@ziccardi Apologies, misunderstanding, will make this change

ziccardi commented on core/src/main/java/org/aerogear/mobile/core/MobileCore.java:
```diff
@@ -174,9 +173,9 @@ public Context getContext() {
      * @return the configuration for this service from the JSON config file
      */
-    private ServiceConfiguration getServiceConfiguration(String type) {
+    private ServiceConfiguration getServiceConfiguration(String type, boolean requiresConfiguration) {
```
Make parameter final

ziccardi commented on example/src/main/java/org/aerogear/mobile/example/ui/SecurityServiceFragment.java:
```diff
@@ -63,22 +68,18 @@
-        view = inflater.inflate(R.layout.fragment_security_service, container, false);
-        ButterKnife.bind(this, view);
-
+    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
```
final parameter

matzew commented on auth/src/main/java/org/aerogear/mobile/security/Check.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.aerogear.mobile.security.checks.RootedCheck;
+
+public enum Check {
```
Check as in "security check" ?

aidenkeating replied:
Yeah, I guess this could be `SecurityCheckType` or something like that

aidenkeating commented:
@ziccardi Addressed comments, mind taking another quick look?

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Override
+    public SecurityCheckExecutor addCheck(SecurityCheckType securityCheckType) {
```
parameters should be final

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,81 @@
+        for (SecurityCheck check : checks) {
+            results.add(check.test(context));
+        }
+        return buildResultArray(results);
```
You know the number of results you are going to get since the beginning (`checks.size()`).
Using an `ArrayList` and then converting it into an array is useless here: use directly an array:

```java
SecurityCheckResult[] results = new SecurityCheckResults[checks.size()];
int i = 0;
for (Security check: checks) {
   results[i++] = check.test(context)
}
```

ziccardi commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    private SecurityCheckResult[] buildResultArray(List<SecurityCheckResult> results) {
+        return results.toArray(new SecurityCheckResult[results.size()]);
```
Why a method for just one line? Better to inline this code.

wojta commented on auth/build.gradle:
```diff
@@ -30,6 +30,8 @@ android {
 dependencies {
     implementation project(path: ':core')
     implementation 'net.openid:appauth'
+    //noinspection GradleCompatible
```
What about this?
`implementation "com.android.support:customtabs:26.1.0"
implementation ('net.openid:appauth', {
       exclude group: 'com.android.support', module: 'customtabs'
})`

aidenkeating replied:
@wojta This seems to work too, would it be possible to get an example of what it does/why it helps? Not too familiar with this whole ecosystem

aidenkeating commented:
@ziccardi One more look when you get the chance?

@wojta Included your suggested change also, thanks 👍

wojta commented on auth/build.gradle:
```diff
@@ -30,8 +30,8 @@ android {
     implementation 'net.openid:appauth'
-    //noinspection GradleCompatible
-    implementation 'com.android.support:customtabs:26.1.0'
+    implementation "com.android.support:customtabs:26.1.0"
```
I'm still actually not sure if this change is not dangerous. It simply replaces dependency, but `appauth` is still developed against 26.1.0.
What if we increase it to 27.0.2? 
```
implementation "com.android.support:customtabs:27.0.2"
    implementation "com.android.support:animated-vector-drawable:27.0.2"
    implementation "com.android.support:appcompat-v7:27.0.2"
    implementation('net.openid:appauth', {
        exclude group: 'com.android.support', module: 'customtabs'
    })
implementation('com.scottyab:rootbeer-lib:0.0.6', {
        exclude group: 'com.android.support', module: 'animated-vector-drawable'
        exclude group: 'com.android.support', module: 'appcompat-v7'
    })    
```
So far I tried it and app works still ok. Can we use it?

TommyJ1994 replied:
@wojta Might work because we have a hardcoded blacklist against the usage of custom tabs - see [here](https://github.com/aerogear/aerogear-android-sdk/blob/38a514725a31723d09038c7e59952cf1c857cae6/auth/src/main/java/org/aerogear/mobile/auth/authenticator/AuthorizationServiceFactory.java#L86-L88) (which needs to change). Have you tried without the blacklist?

aidenkeating replied:
@danielpassos Any opinion on using this?

danielpassos replied:
If we are not going to use this I see no problem

ziccardi commented:
@aidenkeating just a couple of missing 'final' still there from my previous review.

matzew commented on auth/src/main/java/org/aerogear/mobile/security/SecurityCheckType.java:
```diff
@@ -0,0 +1,25 @@
+
+import org.aerogear.mobile.security.checks.RootedCheck;
+
+public enum SecurityCheckType {
```
let's have some javadoc

matzew commented on auth/src/main/java/org/aerogear/mobile/security/SecurityCheckResult.java:
```diff
@@ -0,0 +1,20 @@
+
+import org.aerogear.mobile.core.metrics.Metrics;
+
+public interface SecurityCheckResult {
```
let's have some javadoc

matzew commented on auth/src/main/java/org/aerogear/mobile/security/SecurityCheck.java:
```diff
@@ -0,0 +1,13 @@
+
+import android.content.Context;
+
+public interface SecurityCheck {
```
let's have some javadoc

matzew commented on auth/src/main/java/org/aerogear/mobile/security/SecurityCheckExecutor.java:
```diff
@@ -0,0 +1,35 @@
+
+import org.aerogear.mobile.core.metrics.MetricsService;
+
+public interface SecurityCheckExecutor {
```
let's have some javadoc

matzew commented on auth/src/main/java/org/aerogear/mobile/security/SecurityService.java:
```diff
@@ -0,0 +1,55 @@
+import org.aerogear.mobile.security.impl.SecurityCheckExecutorImpl;
+import org.aerogear.mobile.security.metrics.SecurityCheckResultMetric;
+
+public class SecurityService implements ServiceModule{
```
Service vs. Module ?

aidenkeating replied:
With Auth and Metrics, the two most recent additions we went with Service. So I think it's service now

matzew replied:
a service is not a module :)

matzew commented on auth/src/main/java/org/aerogear/mobile/security/checks/RootedCheck.java:
```diff
@@ -0,0 +1,25 @@
+import org.aerogear.mobile.security.SecurityCheckResult;
+import org.aerogear.mobile.security.impl.SecurityCheckResultImpl;
+
+public class RootedCheck implements SecurityCheck {
```
let's have some class level javadoc

matzew commented on auth/src/main/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorImpl.java:
```diff
@@ -0,0 +1,78 @@
+import java.util.List;
+
+public class SecurityCheckExecutorImpl implements SecurityCheckExecutor {
+    private final Collection<SecurityCheck> checks;
```
why a `Collection`? instead of a `Set`? we can also initialise the thing on this line, instead of in the ctor

ziccardi replied:
I suggested to use a `Collection` as interface and an `HashSet` as implementation.
The reason is that we need to have a set of unique objects, but we don't need any of the methods exposed by a `Set` that are not exposed in a `Collection`. So, I think the most generic interface should be used, so that if we should ever decide to change from HashSet to ArrayList, for example, the rest of the code will stay unchanged.

matzew replied:
fair enough

matzew commented on auth/src/main/java/org/aerogear/mobile/security/metrics/SecurityCheckResultMetric.java:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    private Map<String, String> getDataFromResult(final SecurityCheckResult result) {
+        HashMap<String, String> data = new HashMap<>();
```
use `Map<String, String> data = ...` (and make it `final`)

matzew commented on auth/src/test/java/org/aerogear/mobile/security/impl/SecurityCheckExecutorTest.java:
```diff
@@ -0,0 +1,62 @@
+
+    @Test
+    public void testExecute() {
+        SecurityCheckResult[] results = executor.addCheck(securityCheckType).execute();
```
I think I generally don't understand why we do use that much arrays, instead of a generic collection (e.g. List/Set)

matzew replied:
nothing on this PR - more a general thing :)

aidenkeating commented:
@ziccardi Yep I should probably make sure I addressed the comments before asking for another review 😛 

@ziccardi @matzew Would you mind taking another look?

matzew commented:
@aidenkeating ship it !
